File: src/main/java/frc/robot/subsystems/drive/DriveBase.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems.drive;

import java.util.List;
import java.util.function.DoubleSupplier;
import java.util.function.Supplier;

import org.littletonrobotics.junction.Logger;

import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;

import edu.wpi.first.math.MathUtil;
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.estimator.SwerveDrivePoseEstimator;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Transform2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.geometry.Twist3d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.smartdashboard.Field2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.Constants.DriveConstants;
import frc.robot.Constants.FieldConstants;
import frc.robot.Constants.RobotConstants;
import frc.robot.Constants.DriveConstants.BackLeftModuleConstants;
import frc.robot.Constants.DriveConstants.BackRightModuleConstants;
import frc.robot.Constants.DriveConstants.FrontLeftModuleConstants;
import frc.robot.Constants.DriveConstants.FrontRightModuleConstants;
import frc.robot.Constants.IOConstants.ControllerIOConstants;
import frc.robot.Constants.IOConstants;
import frc.robot.subsystems.vision.Camera;
import frc.robot.subsystems.vision.Camera.Pose;
import frc.robot.util.led.LEDs;
import frc.robot.util.math.LobstahMath;
import frc.robot.util.sysId.CharacterizableSubsystem;
import frc.robot.util.trajectory.AlliancePoseMirror;

public class DriveBase extends CharacterizableSubsystem {
    /** Creates a new SwerveDriveBase. */
    private final SwerveModule[] modules;

    private SwerveDrivePoseEstimator swerveOdometry;
    private SwerveDrivePoseEstimator visionLessOdometry;
    // private SwerveSetpointGenerator setpointGenerator;
    private SwerveSetpoint swerveSetpoint = new SwerveSetpoint(new ChassisSpeeds(), new SwerveModuleState[] {
            new SwerveModuleState(), new SwerveModuleState(), new SwerveModuleState(), new SwerveModuleState() });
    private final GyroIO gyro;
    private final GyroIOInputsAutoLogged gyroInputs = new GyroIOInputsAutoLogged();
    private boolean isOpenLoop;
    private final List<Camera> cameras;
    private boolean hasSeenTag = false;
    private boolean needGyroReset = true;

    private final PIDController xController = new PIDController(DriveConstants.AUTO_ALIGN_TRANSLATION_kP,
            DriveConstants.AUTO_ALIGN_TRANSLATION_kI, DriveConstants.AUTO_ALIGN_TRANSLATION_kD);
    private final PIDController yController = new PIDController(DriveConstants.AUTO_ALIGN_TRANSLATION_kP,
            DriveConstants.AUTO_ALIGN_TRANSLATION_kI, DriveConstants.AUTO_ALIGN_TRANSLATION_kD);
    private final PIDController thetaController = new PIDController(DriveConstants.ROTATION_PID_CONSTANTS.kP,
            DriveConstants.ROTATION_PID_CONSTANTS.kI, DriveConstants.ROTATION_PID_CONSTANTS.kD);

    private Field2d field;

    /**
     * Create a new serve drivebase.
     * 
     * @param gyroIO     The {@link GyroIO}.
     * @param cameras    a list of {@link Camera}s on the robot
     * @param frontLeft
     * @param frontRight
     * @param backLeft
     * @param backRight
     * @param isOpenLoop If true, skip closed-loop (PID) control.
     */
    public DriveBase(GyroIO gyroIO, List<Camera> cameras, SwerveModuleIO frontLeft, SwerveModuleIO frontRight,
            SwerveModuleIO backLeft, SwerveModuleIO backRight, boolean isOpenLoop) {
        this.modules = new SwerveModule[] { new SwerveModule(frontLeft, FrontLeftModuleConstants.moduleID),
                new SwerveModule(frontRight, FrontRightModuleConstants.moduleID),
                new SwerveModule(backLeft, BackLeftModuleConstants.moduleID),
                new SwerveModule(backRight, BackRightModuleConstants.moduleID) };

        this.gyro = gyroIO;

        this.cameras = cameras;
        swerveOdometry = new SwerveDrivePoseEstimator(DriveConstants.KINEMATICS, gyroInputs.yawPosition, getPositions(),
                new Pose2d());
        visionLessOdometry = new SwerveDrivePoseEstimator(DriveConstants.KINEMATICS, gyroInputs.yawPosition,
                getPositions(), new Pose2d());
        // setpointGenerator = new SwerveSetpointGenerator(DriveConstants.KINEMATICS, DriveConstants.MODULE_LOCATIONS);

        field = new Field2d();
        SmartDashboard.putData("Field", field);

        xController.setTolerance(0.02);
        yController.setTolerance(0.02);
        thetaController.enableContinuousInput(-Math.PI, Math.PI);

        this.isOpenLoop = isOpenLoop;
        this.resetPose(getPose());
    }

    /**
     * Resets pose of odometry to a given pose.
     * 
     * @param pose The desired pose to reset the odometry to.
     */
    public void resetPose(Pose2d pose) {
        swerveOdometry.resetPosition(gyroInputs.yawPosition, getPositions(), pose);
        visionLessOdometry.resetPosition(gyroInputs.yawPosition, getPositions(), pose);
    }

    /**
     * Gets pose from odometry.
     * 
     * @return The current estimated pose of the odometry
     */
    public Pose2d getPose() {
        return swerveOdometry.getEstimatedPosition();
    }

    /**
     * Gets states of the four swerve modules.
     * 
     * @return The states of the four swerve modules in a {@link SwerveModuleState}
     *         array.
     */
    public SwerveModuleState[] getStates() {
        SwerveModuleState[] states = new SwerveModuleState[4];
        for (SwerveModule module : modules) {
            states[module.getModuleID()] = module.getState();
        }
        return states;
    }

    /**
     * Gets positions of the four swerve modules.
     * 
     * @return The positions of the four swerve modules in a
     *         {@link SwerveModulePosition} array.
     */
    public SwerveModulePosition[] getPositions() {
        SwerveModulePosition[] positions = new SwerveModulePosition[4];
        for (SwerveModule module : modules) {
            positions[module.getModuleID()] = module.getPosition();
        }
        return positions;
    }

    /**
     * Gets robot relative ChassisSpeeds.
     * 
     * @return The robot-relative {@link ChassisSpeeds}.
     */
    public ChassisSpeeds getRobotRelativeSpeeds() {
        return DriveConstants.KINEMATICS.toChassisSpeeds(getStates());
    }

    /**
     * Drives the robot robot-relative according to provided {@link ChassisSpeeds}.
     * 
     * @param chassisSpeeds The desired ChassisSpeeds. Should be robot relative.
     */
    public void driveRobotRelative(ChassisSpeeds chassisSpeeds) {
        var moduleStates = DriveConstants.KINEMATICS.toSwerveModuleStates(chassisSpeeds);
        Logger.recordOutput("Unoptimized:", moduleStates);
        // swerveSetpoint = setpointGenerator.generateSetpoint(DriveConstants.MODULE_LIMITS,
        //     new SwerveSetpoint(getRobotRelativeSpeeds(), getStates()), ChassisSpeeds.discretize(chassisSpeeds, 0.02),
        //     SimConstants.LOOP_TIME); 
        SwerveDriveKinematics.desaturateWheelSpeeds(moduleStates, DriveConstants.MAX_DRIVE_SPEED);
        setModuleStates(moduleStates);
    }

    /**
     * Sets desired {@link SwerveModuleState}s. Optimizes states.
     * 
     * @param desiredStates The states to set for each module.
     * @return The optimized SwerveModuleStates, now desired states.
     */
    private SwerveModuleState[] setModuleStates(SwerveModuleState[] desiredStates) {
        SwerveModuleState[] optimizedStates = new SwerveModuleState[4];
        // SwerveDriveKinematics.desaturateWheelSpeeds(
        // desiredStates, DriveConstants.MAX_DRIVE_SPEED);
        for (SwerveModule module : modules) {
            optimizedStates[module.getModuleID()] = module.setDesiredState(desiredStates[module.getModuleID()],
                    isOpenLoop);
        }
        swerveSetpoint.moduleStates = optimizedStates;
        Logger.recordOutput("SwerveStates/Desired", desiredStates);
        Logger.recordOutput("SwerveStates/Optimized", swerveSetpoint.moduleStates);
        Logger.recordOutput("SwerveStates/SetpointSpeeds", swerveSetpoint.chassisSpeeds);
        return desiredStates;
    }

    /**
     * Converts robot relative {@link ChassisSpeeds} to field relative.
     * 
     * @param robotRelativeSpeeds The robot relative speeds to convert
     * @return The field relative ChassisSpeeds.
     */
    public ChassisSpeeds getFieldRelativeChassisSpeeds(ChassisSpeeds robotRelativeSpeeds) {
        Rotation2d angle = getPose().getRotation();
        return new ChassisSpeeds(
                robotRelativeSpeeds.vxMetersPerSecond * angle.getCos()
                        - robotRelativeSpeeds.vyMetersPerSecond * angle.getSin(),
                robotRelativeSpeeds.vyMetersPerSecond * angle.getCos()
                        + robotRelativeSpeeds.vxMetersPerSecond * angle.getSin(),
                robotRelativeSpeeds.omegaRadiansPerSecond);
    }

    /**
     * Sets the {@link IdleMode} of the DriveBase motors.
     * 
     * @param mode The braking mode (Coast or Brake) of the swerve module motors.
     */
    public void setIdleMode(IdleMode mode) {
        for (SwerveModule module : modules) {
            module.setIdleMode(mode);
        }
    }

    /** Stops all of the modules' motors. */
    private void stopMotors() {
        for (SwerveModule module : modules) {
            module.stop();
        }
    }

    /**
     * @return Whether or not the controller is open loop.
     */
    public boolean isOpenLoop() {
        return isOpenLoop;
    }

    /**
     * Sets whether the controller is open loop.
     * 
     * @param newValue The new boolean to set.
     */
    public void setIsOpenLoop(boolean newValue) {
        isOpenLoop = newValue;
    }

    /**
     * Gets gyro angle
     * 
     * @return The angle of the gyro as a {@link Rotation2d}.
     */
    public Rotation2d getGyroAngle() {
        return gyroInputs.yawPosition;
    }

    @Override
    /**
     * Runs motors during characterization voltage ramp routines.
     * 
     * @param Voltage to run motors at.
     * @see CharacterizableSubsystem
     */
    public void runVolts(double volts) {
        for (SwerveModule module : modules) {
            module.runVolts(volts);
        }
    }

    @Override
    public void periodic() {

        if (needGyroReset && !gyroInputs.isCalibrating) {
            gyro.zeroGyro();
            needGyroReset = false;
        }

        swerveOdometry.updateWithTime(Timer.getFPGATimestamp(), gyroInputs.yawPosition, getPositions());
        visionLessOdometry.updateWithTime(Timer.getFPGATimestamp(), gyroInputs.yawPosition, getPositions());
        SmartDashboard.putBoolean("Has seen tag", hasSeenTag);
        for (Camera camera : cameras) {
            camera.periodic();
            Pose estimatedPose = camera.getEstimatedPose(getPose());
            if (estimatedPose.pose().isPresent()
                    && (hasSeenTag == false
                            || LobstahMath.getDistBetweenPoses(estimatedPose.pose().get().toPose2d(), getPose()) <= 8)
                    && Math.abs(estimatedPose.pose().get().getZ()) < 0.1 && (camera.getName().startsWith("front"))) {
                if (hasSeenTag == false) {
                    resetPose(new Pose2d(estimatedPose.pose().get().getX(), estimatedPose.pose().get().getY(),
                            getGyroAngle()));
                    hasSeenTag = true;
                }
                swerveOdometry.addVisionMeasurement(estimatedPose.pose().get().toPose2d(),
                        estimatedPose.timestamp().get(), estimatedPose.stdev().get());
                Logger.recordOutput("Vision/" + camera.getName() + "Used", true);
            } else
                Logger.recordOutput("Vision/" + camera.getName() + "Used", false);
        }
        resetPose(new Pose2d(MathUtil.clamp(getPose().getX(), 0, FieldConstants.FIELD_LENGTH),
                MathUtil.clamp(getPose().getY(), 0, FieldConstants.FIELD_WIDTH), getPose().getRotation()));
        field.setRobotPose(getPose());
        Logger.recordOutput("Odometry", getPose());
        Logger.recordOutput("Vision Less", visionLessOdometry.getEstimatedPosition());
        gyro.updateInputs(gyroInputs);
        Logger.processInputs("Drive/Gyro", gyroInputs);
        // Log 3D odometry pose
        Pose3d robotPose3d = new Pose3d(getPose());
        robotPose3d = robotPose3d
                .exp(new Twist3d(0.0, 0.0,
                        Math.abs(gyroInputs.pitchPosition.getRadians()) * RobotConstants.TRACK_WIDTH / 2.0, 0.0,
                        gyroInputs.pitchPosition.getRadians(), 0.0))
                .exp(new Twist3d(0.0, 0.0,
                        Math.abs(gyroInputs.rollPosition.getRadians()) * RobotConstants.TRACK_WIDTH / 2.0,
                        gyroInputs.rollPosition.getRadians(), 0.0, 0.0));

        Logger.recordOutput("Odometry/Robot3d", robotPose3d);
        for (Camera camera : cameras) {
            Logger.recordOutput("Vision/" + camera.getName() + "/Pose3d", robotPose3d.plus(camera.getRobotToCamera()));
        }
        SmartDashboard.putBoolean("Field Centric", DriveConstants.FIELD_CENTRIC);
        for (var module : modules) {
            module.periodic();
        }
        if (DriverStation.isDisabled()) {
            // Stop moving while disabled
            for (var module : modules) {
                module.stop();
            }
        }

        Logger.recordOutput("SwerveStates/Measured", getStates());
        SmartDashboard.putData("Drivebase subsystem", this);
    }

    /**
     * Constructs a command which holds this in place.
     * 
     * @return the constructed command
     */
    public Command stop() {
        return run(this::stopMotors);
    }

    /**
     * Constructs a command which stops this and immediately returns.
     * 
     * @return the constructed command
     */
    public Command stopOnce() {
        return runOnce(this::stopMotors);
    }

    /**
     * Constructs a command to auto-align to a particular pose. It will also light
     * the aligning LEDs at the beginning and turn them off at thend.
     * 
     * @param targetPoseSupplier a supplier which supplies the pose; the target pose
     *                           will be obtained from this supplier every time the
     *                           command is scheduled.
     * @return the constructed command
     */
    public Command alignToPose(Supplier<Pose2d> targetPoseSupplier) {
        return startRun(() -> {
            xController.reset();
            yController.reset();
            thetaController.reset();
            Pose2d targetPose = targetPoseSupplier.get();
            Logger.recordOutput("AutoAlignTargetPose", targetPose);
            xController.setSetpoint(targetPose.getX());
            yController.setSetpoint(targetPose.getY());
            thetaController.setSetpoint(targetPose.getRotation().getRadians());
        }, () -> {
            driveRobotRelative(ChassisSpeeds.fromFieldRelativeSpeeds(xController.calculate(getPose().getX()),
                    yController.calculate(getPose().getY()),
                    thetaController.calculate(getPose().getRotation().getRadians()), getPose().getRotation()));
        }).until(() -> xController.atSetpoint() && yController.atSetpoint() && thetaController.atSetpoint())
                .andThen(stop()).beforeStarting(() -> LEDs.getInstance().setAligning(true))
                .finallyDo(() -> LEDs.getInstance().setAligning(false));
    }

    /**
     * Create a new command to drive field-relative.
     * 
     * @param strafeXSupplier  Supplier for strafe in X direction, e.g. from a
     *                         joystick.
     * @param strafeYSupplier  Supplier for strafe in Y direction, e.g. from a
     *                         joystick.
     * @param rotationSupplier Supplier for rotation, e.g. from a joystick.
     * @return constructed command
     */
    public Command joystickDrive(DoubleSupplier strafeXSupplier, DoubleSupplier strafeYSupplier,
            DoubleSupplier rotationSupplier) {
        return run(() -> {
            double linearMagnitude = MathUtil.applyDeadband(
                    Math.hypot(strafeXSupplier.getAsDouble(), strafeYSupplier.getAsDouble()),
                    IOConstants.JOYSTICK_DEADBAND);
            Rotation2d linearDirection = linearMagnitude > 0
                    ? new Rotation2d(strafeXSupplier.getAsDouble(), strafeYSupplier.getAsDouble())
                    : Rotation2d.kZero;
            if (AlliancePoseMirror.isRedAlliance()) linearDirection = linearDirection.plus(Rotation2d.k180deg);
            double omega = MathUtil.applyDeadband(rotationSupplier.getAsDouble(), IOConstants.JOYSTICK_DEADBAND);

            // Square values
            if (ControllerIOConstants.SQUARE_INPUTS) {
                linearMagnitude = linearMagnitude * linearMagnitude;
                omega = Math.copySign(omega * omega, omega);
            }

            // Calculate new linear velocity
            Translation2d linearVelocity = new Pose2d(new Translation2d(), linearDirection)
                    .transformBy(new Transform2d(linearMagnitude, 0.0, new Rotation2d())).getTranslation();

            ChassisSpeeds chassisSpeeds = new ChassisSpeeds(linearVelocity.getX() * DriveConstants.MAX_DRIVE_SPEED,
                    linearVelocity.getY() * DriveConstants.MAX_DRIVE_SPEED, omega * DriveConstants.MAX_ANGULAR_SPEED);

            driveRobotRelative(ChassisSpeeds.fromFieldRelativeSpeeds(chassisSpeeds, getPose().getRotation()));
        }).finallyDo(this::stopMotors);
    }

    /**
     * Drive the robot robot-relative.
     * 
     * @param strafeX  constant strafe in X (m/s)
     * @param strafeY  constant strafe in Y (m/s)
     * @param rotation constant rotation (rad/s)
     * @return constructed command
     */
    public Command relativeDrive(double strafeX, double strafeY, double rotation) {
        ChassisSpeeds speeds = new ChassisSpeeds(strafeX, strafeY, rotation);
        return run(() -> {
            driveRobotRelative(speeds);
        }).finallyDo(this::stopMotors);
    }
}

File: src/main/java/frc/robot/subsystems/drive/DriveBase.kt
package frc.robot.subsystems.drive

import com.revrobotics.spark.config.SparkBaseConfig
import edu.wpi.first.math.MathUtil
import edu.wpi.first.math.controller.PIDController
import edu.wpi.first.math.estimator.SwerveDrivePoseEstimator
import edu.wpi.first.math.geometry.*
import edu.wpi.first.math.kinematics.ChassisSpeeds
import edu.wpi.first.math.kinematics.SwerveDriveKinematics
import edu.wpi.first.math.kinematics.SwerveModulePosition
import edu.wpi.first.math.kinematics.SwerveModuleState
import edu.wpi.first.wpilibj.DriverStation
import edu.wpi.first.wpilibj.Timer
import edu.wpi.first.wpilibj.smartdashboard.Field2d
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import edu.wpi.first.wpilibj2.command.Command
import frc.robot.Constants.*
import frc.robot.Constants.IOConstants.ControllerIOConstants
import frc.robot.FieldConstants
import frc.robot.subsystems.vision.Camera
import frc.robot.util.led.LEDs
import frc.robot.util.math.LobstahMath
import frc.robot.util.sysId.CharacterizableSubsystem
import frc.robot.util.trajectory.AlliancePoseMirror
import org.littletonrobotics.junction.Logger
import java.util.function.DoubleSupplier
import java.util.function.Supplier
import kotlin.math.abs
import kotlin.math.hypot
import kotlin.math.withSign

class DriveBase(
    private val gyro: GyroIO,
    private val cameras: List<Camera>,
    frontLeft: SwerveModuleIO,
    frontRight: SwerveModuleIO,
    backLeft: SwerveModuleIO,
    backRight: SwerveModuleIO,
    private var isOpenLoop: Boolean
) : CharacterizableSubsystem() {
    private val modules = listOf(
        SwerveModule(frontLeft, DriveConstants.FrontLeftModuleConstants.moduleID),
        SwerveModule(frontRight, DriveConstants.FrontRightModuleConstants.moduleID),
        SwerveModule(backLeft, DriveConstants.BackLeftModuleConstants.moduleID),
        SwerveModule(backRight, DriveConstants.BackRightModuleConstants.moduleID)
    )

    // private val setpointGenerator = SwerveSetpointGenerator(DriveConstants.KINEMATICS, DriveConstants.MODULE_LOCATIONS)
    // private var swerveSetpoint = SwerveSetpoint(ChassisSpeeds(), Array(4) { SwerveModuleState())
    private val gyroInputs = GyroIOInputsAutoLogged()
    private var hasSeenTag = false
    private var needGyroReset = true
    private val swerveOdometry = SwerveDrivePoseEstimator(
        DriveConstants.KINEMATICS, gyroInputs.yawPosition, positions, Pose2d.kZero
    )
    private val visionlessOdometry = SwerveDrivePoseEstimator(
        DriveConstants.KINEMATICS, gyroInputs.yawPosition, positions, Pose2d.kZero
    )

    private val xController = PIDController(
        DriveConstants.AUTO_ALIGN_TRANSLATION_kP,
        DriveConstants.AUTO_ALIGN_TRANSLATION_kI,
        DriveConstants.AUTO_ALIGN_TRANSLATION_kD
    )
    private val yController = PIDController(
        DriveConstants.AUTO_ALIGN_TRANSLATION_kP,
        DriveConstants.AUTO_ALIGN_TRANSLATION_kI,
        DriveConstants.AUTO_ALIGN_TRANSLATION_kD
    )
    private val thetaController = PIDController(
        DriveConstants.ROTATION_PID_CONSTANTS.kP,
        DriveConstants.ROTATION_PID_CONSTANTS.kI,
        DriveConstants.ROTATION_PID_CONSTANTS.kD
    )
    private val field = Field2d()

    init {
        SmartDashboard.putData("Field", field)
        xController.setTolerance(0.02)
        yController.setTolerance(0.02)
        thetaController.enableContinuousInput(-Math.PI, Math.PI)
        resetPose(pose)
    }

    /**
     * Resets pose of odometry to a given pose.
     *
     * @param pose The desired pose to reset the odometry to.
     */
    fun resetPose(pose: Pose2d) {
        val newPose = Pose2d(pose.translation, gyroInputs.yawPosition)
        swerveOdometry.resetPosition(gyroInputs.yawPosition, positions, newPose)
        visionlessOdometry.resetPosition(gyroInputs.yawPosition, positions, newPose)
    }

    /**
     * The current estimated odometry pose. Setting this property resets odometry using [resetPose].
     *
     * @see SwerveDrivePoseEstimator
     */
    var pose: Pose2d
        get() = swerveOdometry.estimatedPosition
        set(value) = resetPose(value)

    /**
     * The states of the four swerve modules. Setting this sets the desired swerve module states.
     *
     * @see SwerveModuleState
     */
    var states: Array<SwerveModuleState>
        get() = modules.sortedBy { it.moduleID }.map { it.state }.toTypedArray()
        set(desiredStates) {
            val optimizedStates = arrayOfNulls<SwerveModuleState>(4)


            // SwerveDriveKinematics.desaturateWheelSpeeds(
            // desiredStates, DriveConstants.MAX_DRIVE_SPEED);
            for (module in modules) {
                optimizedStates[module.moduleID] = module.setDesiredState(
                    desiredStates[module.moduleID], isOpenLoop
                )
            }
            // swerveSetpoint.moduleStates = optimizedStates
            Logger.recordOutput("SwerveStates/Desired", *desiredStates)
            Logger.recordOutput("SwerveStates/Optimized", *optimizedStates)
        }

    /**
     * The positions of the four swerve modules.
     *
     * @see SwerveModulePosition
     */
    val positions: Array<SwerveModulePosition>
        get() = modules.sortedBy { it.moduleID }.map { it.position }.toTypedArray()

    /**
     * The robot-relative chassis speeds. Setting this drives the robot robot-relative according to the provided speeds.
     */
    var robotRelativeSpeeds: ChassisSpeeds
        get() = DriveConstants.KINEMATICS.toChassisSpeeds(*states)
        set(chassisSpeeds) {
            val desiredStates = DriveConstants.KINEMATICS.toSwerveModuleStates(chassisSpeeds)
            Logger.recordOutput("Unoptimized:", *desiredStates)
            SwerveDriveKinematics.desaturateWheelSpeeds(desiredStates, DriveConstants.MAX_DRIVE_SPEED)
            states = desiredStates
        }

    /**
     * The field relative chassis speeds of the robot.
     */
    val fieldRelativeSpeeds: ChassisSpeeds
        get() = pose.rotation.let { angle ->
            ChassisSpeeds(
                robotRelativeSpeeds.vxMetersPerSecond * angle.cos - robotRelativeSpeeds.vyMetersPerSecond * angle.sin,
                robotRelativeSpeeds.vyMetersPerSecond * angle.cos + robotRelativeSpeeds.vxMetersPerSecond * angle.sin,
                robotRelativeSpeeds.omegaRadiansPerSecond
            )
        }

    var idleMode: SparkBaseConfig.IdleMode = SparkBaseConfig.IdleMode.kBrake
        set(value) {
            field = value
            modules.forEach { it.setIdleMode(field) }
        }

    /**
     * Stop all the modules' motors.
     */
    private fun stopMotors() = modules.forEach { it.stop() }

    val gyroAngle: Rotation2d
        get() = gyroInputs.yawPosition

    /**
     * Runs motors during characterization voltage ramp routines.
     *
     * @param volts Voltage to run motors at.
     * @see CharacterizableSubsystem
     */
    override fun runVolts(volts: Double) = modules.forEach { it.runVolts(volts) }

    override fun periodic() {
        if (needGyroReset and !gyroInputs.isCalibrating) {
            gyro.zeroGyro()
            needGyroReset = false
        }

        swerveOdometry.updateWithTime(Timer.getFPGATimestamp(), gyroInputs.yawPosition, positions)
        visionlessOdometry.updateWithTime(Timer.getFPGATimestamp(), gyroInputs.yawPosition, positions)
        Logger.recordOutput("Has seen tag", hasSeenTag)
        cameras.forEach { camera ->
            camera.periodic()
            val estimatedPose = camera.getEstimatedPose(pose)
            if (estimatedPose.pose.isPresent and (!hasSeenTag or (LobstahMath.getDistBetweenPoses(
                    estimatedPose.pose.get(), pose
                ) <= 8)) and (abs(
                    estimatedPose.pose.get().z
                ) < 0.3)
            ) {
                if (!hasSeenTag) {
                    pose = estimatedPose.pose.get().toPose2d()
                    hasSeenTag = true
                }
                swerveOdometry.addVisionMeasurement(
                    estimatedPose.pose.get().toPose2d(), estimatedPose.timestamp.get(), estimatedPose.stdev().get()
                )
                Logger.recordOutput("Vision/" + camera.name + "Used", true)
            } else Logger.recordOutput("Vision/" + camera.name + "Used", false)
        }
        pose = Pose2d(
            MathUtil.clamp(pose.x, 0.0, FieldConstants.fieldLength),
            MathUtil.clamp(pose.y, 0.0, FieldConstants.fieldWidth),
            pose.rotation
        )
        field.robotPose = pose
        Logger.recordOutput("Odometry", pose)
        Logger.recordOutput("Visionless", visionlessOdometry.estimatedPosition)
        gyro.updateInputs(gyroInputs)
        Logger.processInputs("Drive/Gyro", gyroInputs)

        val robotPose3d = Pose3d(pose).exp(
            Twist3d(
                0.0,
                0.0,
                abs(gyroInputs.pitchPosition.radians) * RobotConstants.TRACK_WIDTH / 2.0,
                0.0,
                gyroInputs.pitchPosition.radians,
                0.0
            )
        ).exp(
                Twist3d(
                    0.0,
                    0.0,
                    abs(gyroInputs.rollPosition.radians) * RobotConstants.TRACK_WIDTH / 2.0,
                    gyroInputs.rollPosition.radians,
                    0.0,
                    0.0
                )
            )
        Logger.recordOutput("Odometry/Robot3d", robotPose3d)
        cameras.forEach { camera ->
            Logger.recordOutput("Vision/" + camera.name + "/Pose3d", robotPose3d.plus(camera.robotToCamera))
        }
        modules.forEach { it.periodic() }
        if (DriverStation.isDisabled()) stopMotors()

        Logger.recordOutput("SwerveStates/Measured", *states)
        SmartDashboard.putData("Drivebase subsystem", this)
    }

    /**
     * Constructs a command which holds this in place.
     *
     * @return the constructed command
     */
    fun stop(): Command = run(this::stopMotors)

    /**
     * Constructs a command which stops this and immediately returns.
     *
     * @return the constructed command
     */
    fun stopOnce(): Command = runOnce(this::stopMotors)

    /**
     * Constructs a command to auto-align to a particular pose. It will also light
     * the aligning LEDs at the beginning and turn them off at the end.
     *
     * @param targetPoseSupplier a supplier which supplies the pose; the target pose
     * will be obtained from this supplier every time the
     * command is scheduled.
     * @return the constructed command
     */
    fun alignToPose(targetPoseSupplier: Supplier<Pose2d>): Command = startRun({
        xController.reset()
        yController.reset()
        thetaController.reset()
        val targetPose = targetPoseSupplier.get()
        Logger.recordOutput("AutoAlignTargetPose", targetPose)
        xController.setSetpoint(targetPose.x)
        yController.setSetpoint(targetPose.y)
        thetaController.setSetpoint(targetPose.rotation.radians)
    }, {
        robotRelativeSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(
            xController.calculate(pose.x),
            yController.calculate(pose.y),
            thetaController.calculate(pose.rotation.radians),
            pose.rotation
        )

    }).until { xController.atSetpoint() && yController.atSetpoint() && thetaController.atSetpoint() }.andThen(stop())
        .beforeStarting({ LEDs.getInstance().setAligning(true) })
        .finallyDo { _ -> LEDs.getInstance().setAligning(false) }


    /**
     * Create a new command to drive field-relative.
     *
     * @param strafeXSupplier  Supplier for strafe in X direction, e.g. from a
     * joystick.
     * @param strafeYSupplier  Supplier for strafe in Y direction, e.g. from a
     * joystick.
     * @param rotationSupplier Supplier for rotation, e.g. from a joystick.
     * @param assist           Whether to apply drive assist.
     * @return constructed command
     */
    @JvmOverloads
    fun joystickDrive(
        strafeXSupplier: DoubleSupplier,
        strafeYSupplier: DoubleSupplier,
        rotationSupplier: DoubleSupplier,
        assist: Boolean = false
    ): Command = run {
        var linearMagnitude = MathUtil.applyDeadband(
            hypot(strafeXSupplier.asDouble, strafeYSupplier.asDouble), IOConstants.JOYSTICK_DEADBAND
        )
        var linearDirection = if (linearMagnitude > 0) Rotation2d(strafeXSupplier.asDouble, strafeYSupplier.asDouble)
        else Rotation2d.kZero
        if (AlliancePoseMirror.isRedAlliance()) linearDirection = linearDirection.plus(Rotation2d.k180deg)
        var omega = MathUtil.applyDeadband(rotationSupplier.asDouble, IOConstants.JOYSTICK_DEADBAND)

        // Square values
        if (ControllerIOConstants.SQUARE_INPUTS) {
            linearMagnitude *= linearMagnitude
            omega = (omega * omega).withSign(omega)
        }

        // Calculate new linear velocity
        val linearVelocity = Pose2d(Translation2d(), linearDirection).transformBy(
                Transform2d(
                    linearMagnitude,
                    0.0,
                    Rotation2d()
                )
            ).translation

        if (assist && AlliancePoseMirror.mirrorPose2d(pose).x < FieldConstants.LinesVertical.neutralZoneNear) {
            val targetPose =
                AlliancePoseMirror.mirrorTranslation2d(FieldConstants.Hub.innerCenterPoint.toTranslation2d())
            val targetAngle = targetPose.minus(pose.translation).angle
            omega = thetaController.calculate(
                pose.rotation.plus(Rotation2d.kCCW_Pi_2).radians, targetAngle.radians
            )
        }

        val chassisSpeeds = ChassisSpeeds(
            linearVelocity.x * DriveConstants.MAX_DRIVE_SPEED,
            linearVelocity.y * DriveConstants.MAX_DRIVE_SPEED,
            omega * DriveConstants.MAX_ANGULAR_SPEED
        )
        robotRelativeSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(chassisSpeeds, pose.rotation)
    }.finallyDo(this::stopMotors)

    /**
     * Drive the robot robot-relative.
     *
     * @param strafeX  constant strafe in X (m/s)
     * @param strafeY  constant strafe in Y (m/s)
     * @param rotation constant rotation (rad/s)
     * @return constructed command
     */
    fun relativeDrive(strafeX: Double, strafeY: Double, rotation: Double): Command {
        val speeds = ChassisSpeeds(strafeX, strafeX, rotation)
        return run { robotRelativeSpeeds = speeds }.finallyDo(this::stopMotors)
    }
}